File: controllers/rewardController.js
// =====================================================
// Controller: Khen thưởng & Kỷ luật
// =====================================================
const RewardModel = require('../models/rewardModel');

const getAll = async (req, res) => {
    try {
        const options = {
            page: parseInt(req.query.page) || 1,
            limit: parseInt(req.query.limit) || 20,
            maNhanVien: req.query.maNhanVien || '',
            loai: req.query.loai || '',
            search: req.query.search || ''
        };
        const [rows, total] = await Promise.all([
            RewardModel.getAll(options),
            RewardModel.count(options)
        ]);
        return res.json({ success: true, data: rows, pagination: { page: options.page, limit: options.limit, total, totalPages: Math.ceil(total / options.limit) } });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi lấy danh sách' });
    }
};

const getOne = async (req, res) => {
    try {
        const item = await RewardModel.findById(req.params.id);
        if (!item) return res.status(404).json({ success: false, message: 'Không tìm thấy' });
        return res.json({ success: true, data: item });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi máy chủ' });
    }
};

const create = async (req, res) => {
    try {
        if (!req.body.maNhanVien || !req.body.loai || !req.body.ngay) {
            return res.status(400).json({ success: false, message: 'Thiếu thông tin bắt buộc' });
        }
        const id = await RewardModel.create(req.body);
        const item = await RewardModel.findById(id);
        return res.status(201).json({ success: true, message: 'Thêm thành công', data: item });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi thêm: ' + error.message });
    }
};

const update = async (req, res) => {
    try {
        await RewardModel.update(req.params.id, req.body);
        const item = await RewardModel.findById(req.params.id);
        return res.json({ success: true, message: 'Cập nhật thành công', data: item });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi cập nhật' });
    }
};

const remove = async (req, res) => {
    try {
        await RewardModel.delete(req.params.id);
        return res.json({ success: true, message: 'Xóa thành công' });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi xóa' });
    }
};

module.exports = { getAll, getOne, create, update, remove };

File: quan.comdemo/quan.comdemo/src/main/java/K23/CNT1/_5/DAQ/quan/comdemo/controller/RewardController.java
package K23.CNT1._5.DAQ.quan.comdemo.controller;

import K23.CNT1._5.DAQ.quan.comdemo.dto.ApiResponse;
import K23.CNT1._5.DAQ.quan.comdemo.entity.KhenThuongKyLuat;
import K23.CNT1._5.DAQ.quan.comdemo.service.RewardService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/rewards")
public class RewardController {

    @Autowired
    private RewardService rewardService;

    @GetMapping
    public ResponseEntity<ApiResponse<List<KhenThuongKyLuat>>> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit,
            @RequestParam(required = false) Integer maNhanVien,
            @RequestParam(required = false, defaultValue = "") String loai,
            @RequestParam(required = false, defaultValue = "") String search) {

        Page<KhenThuongKyLuat> resultPage = rewardService.getAll(page, limit, maNhanVien, loai, search);

        Map<String, Object> pagination = new HashMap<>();
        pagination.put("page", page);
        pagination.put("limit", limit);
        pagination.put("total", resultPage.getTotalElements());
        pagination.put("totalPages", resultPage.getTotalPages());

        return ResponseEntity.ok(ApiResponse.ok(resultPage.getContent(), pagination));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<KhenThuongKyLuat>> getOne(@PathVariable Integer id) {
        KhenThuongKyLuat item = rewardService.getOne(id);
        if (item == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ApiResponse.error("Không tìm thấy"));
        }
        return ResponseEntity.ok(ApiResponse.ok(item));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<KhenThuongKyLuat>> create(@RequestBody Map<String, Object> body) {
        try {
            KhenThuongKyLuat created = rewardService.create(body);
            return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.ok("Thêm thành công", created));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi thêm: " + e.getMessage()));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<KhenThuongKyLuat>> update(@PathVariable Integer id, @RequestBody Map<String, Object> body) {
        try {
            KhenThuongKyLuat updated = rewardService.update(id, body);
            return ResponseEntity.ok(ApiResponse.ok("Cập nhật thành công", updated));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi cập nhật: " + e.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<String>> delete(@PathVariable Integer id) {
        try {
            rewardService.delete(id);
            return ResponseEntity.ok(ApiResponse.ok("Xóa thành công", null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi xóa: " + e.getMessage()));
        }
    }
}
